### PR TITLE
impl disruptor based pojo partition

### PR DESCRIPTION
**Pull request by iamsmkr**
### What changes were proposed in this pull request?
Disruptor pattern implementation of PojoBasedPartition

### Why are the changes needed?
PojoBasedPartition implementation was single threaded. We have changed it to disruptor based multi-threaded implementation.

### Does this PR introduce any user-facing change? If yes is this documented?
No

### How was this patch tested?
Unit, Integration tests

### Are there any further changes required?
No

### CODE REVIEW

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    arrOfMapOfVertices(Math.abs(vId % N_LOAD_THREADS).toInt)
+
+  private def checkIfAlreadyAdded(vId: Long): Unit                        =
+    while (!getVerticesMap(vId).isDefinedAt(vId))
+      Thread.`yield`()
+
+  class VertexAddEventHandler extends EventHandler[VertexAddEvent] {
+
+    override def onEvent(event: VertexAddEvent, sequence: Long, endOfBatch: Boolean): Unit =
+      if (event.dstId == -1L) addVertex(event) else addEdge(event)
```
[the remark this answers is not in the document]
Comment (miratepuffin) [reply core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala:69]: Ok don't do this, but please add comments

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+    private def addVertex(event: VertexAddEvent): Unit = {
+      val VertexAddEvent(msgTime, index, srcId, _, properties, vertexType, _, _) = event
+
+      val vertices = getVerticesMap(srcId)
+      vertices.get(srcId) match { // Check if the vertex exists
+        case Some(vertex) => // If it does
+          vertex revive (msgTime, index) // Add the history point
+          addProperties(msgTime, index, vertex, properties)
+          logger.trace(s"History point added to vertex: $srcId")
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala:82]: Combine these and the one below.

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+          logger.trace(s"Properties added: $properties")
+          logger.trace(s"New vertex created $srcId")
+      }
+    }
+
+    private def addEdge(event: VertexAddEvent): Unit =
+      event.edgeToCreate.getOrElse(
+              new Exception(s"Unspecified type of edge to create. EdgeToCreate = ${event.edgeToCreate}")
+      ) match {
+        case LocalEdge            => addLocalEdge(event)
```
[the remark this answers is not in the document]
Comment (miratepuffin) [reply core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala:99]: c>()->()

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+    private def addIncomingEdge(event: VertexAddEvent): Unit = {
+      val VertexAddEvent(msgTime, index, srcId, dstId, properties, _, edgeType, _) = event
+
+      checkIfAlreadyAdded(dstId)
+
+      val dstVertex = getVerticesMap(dstId)(dstId)
+      dstVertex.getIncomingEdge(srcId) match {
+        case Some(edge) =>
+          edge revive (msgTime, index) // Revive the edge
+          edge.setType(edgeType.map(_.name))
```
[the remark this answers is not in the document]
Comment (miratepuffin) [reply core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala:193]: If you are reviving we do not need to set the type

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+
+      (0 until N_LOAD_THREADS).foreach { i =>
+        if (queues(i).remainingCapacity() != QUEUE_SIZE)
+          finished = false
+      }
+
+      if (!finished) Thread.sleep(10L)
+    }
+
+    (0 until N_LOAD_THREADS).foreach(i => disruptors(i).halt())
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala:268]: We do not need to halt the disruptors -- as per our conversation

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+
+  final case object RemoteIncomingEdge extends EdgeToCreate
+
+  final private[DisruptorQueue] case class VertexAddEvent(
+      private[pojograph] var msgTime: Long = -1L,
+      private[pojograph] var index: Long = -1L,
+      private[pojograph] var srcId: Long = -1L,
+      private[pojograph] var dstId: Long = -1L,
+      private[pojograph] var properties: Properties = Properties(),
+      private[pojograph] var vertexType: Option[Type] = None,
```
Comment (miratepuffin) [inline core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala:293]: I think these two can be merged

**Review by miratepuffin — CHANGES_REQUESTED** [review]

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+      this.vertexType = None
+      this.edgeType = edgeType
+      this.edgeToCreate = Some(edgeToCreate)
+    }
+
+  }
+
+  private[DisruptorQueue] object VertexAddEvent {
+
+    final val EVENT_FACTORY =
```
[the remark this answers is not in the document]
Comment (iamsmkr) [reply core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala:338]: This won't work. Disruptors are based on Fly Weight pattern instead.

File: core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+        case RemoteIncomingEdge   => addIncomingEdge(event)
+        case RemoteOutgoingEdge   => addOutgoingEdge(event)
+      }
+
+    private def addLocalEdge(event: VertexAddEvent): Unit = {
+      val VertexAddEvent(msgTime, index, srcId, dstId, properties, _, edgeType, _) = event
+
+      checkIfAlreadyAdded(srcId)
+
+      val srcVertex = getVerticesMap(srcId)(srcId)
```
[the remark this answers is not in the document]
Comment (iamsmkr) [reply core/src/main/scala/com/raphtory/internals/storage/pojograph/DisruptorQueue.scala:115]: getVerticesMap(srcId) is being used in a lot other places. Combining them would require another method which is more code

**Review by ljeub-pometry — DISMISSED** [review]
Remove the println and turn into a log message and this should be good to merge

**Review by miratepuffin — APPROVED** [review]

**Pull request merged**